### ISSUE TITLE
Try lifting `Session<Action = _>` bounds into `impl` for better errors

### ISSUE THREAD

**Issue by kwf** [boltlabs-inc/dialectic#110]
Currently, the `Action` for methods on `Chan`s is restricted in the `where`-clause of the methods themselves. This means that errors involving session type mismatches restate the full session type of the expected session in a way that can be hard to parse. It's possible that moving this bound upward to the enclosing `impl` block for each method would cause these errors to be easier to read. This change will require each method to be enclosed in its own `impl` block.

Upsides:
- better error messages?

Downsides:
- documentation will become more verbose, because there will be more separate `impl` blocks
- potentially it won't actually improve error messages?

**Comment by yaymukund** [comment]
I do not think this is better 🥲

| Before  | After |
| ------------- | ------------- |
| ![screenshot-2021-05-21-16-05-44](https://user-images.githubusercontent.com/590450/119162413-3b4e3d80-ba52-11eb-8a24-70f53161e52b.png)  | ![screenshot-2021-05-21-16-05-09](https://user-images.githubusercontent.com/590450/119162387-34bfc600-ba52-11eb-804d-c02d3b1b96f4.png)  |

I've pushed my work to https://github.com/boltlabs-inc/dialectic/compare/better-error-experiment in case you wanna doublecheck.

**Comment by kwf** [comment]
Following up on this since we talked about it verbally: I agree with you, sadly. Thanks for trying it out!
